modules/training.py

```diff
@@ -245,11 +247,38 @@ def do_train(lora_name: str, always_override: bool, save_steps: int, micro_batch
     shared.tokenizer.pad_token_id = 0
     shared.tokenizer.padding_side = "left"
 
+    def encode(text, add_bos_token):
+        result = shared.tokenizer.encode(text, truncation=True, max_length=cutoff_len)
+        if not add_bos_token and result[0] == shared.tokenizer.bos_token_id:
+            result = result[1:]
+        return result
+
     def tokenize(prompt):
-        result = shared.tokenizer(prompt, truncation=True, max_length=cutoff_len + 1, padding="max_length")
+
+        if train_only_after == '' or train_only_after not in prompt:
+            input_ids = encode(prompt, True)
+            input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
+            labels = [1] * len(input_ids)
+
+        else:
+            ind = prompt.index(train_only_after) + len(train_only_after)
+            before_tokens = encode(prompt[:ind], False)
```

Comment:
Just one more question, why remove the bos token from before _tokens?

Reply:
Wait you're right good question. I think the answer is I need proper sleep before submitting PRs.
That probably shouldn't be stripped.

Reply:
Heh. I have added it back here: https://github.com/oobabooga/text-generation-webui/commit/63ce5f9c28829a24be2d0eb3f2e57c222c14c042

(if you later conclude that no, we don't want that bos token, let me know)